src/compiler/sys.ts

```diff
@@ -293,6 +293,50 @@ namespace ts {
                     removeFile: removeFile
                 };
             }
+            
+            function createWatchedFileSet() {
+                let watchedDirectories: { [path: string]: FileWatcher } = {};
+                let watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+                
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    let watchedPaths = Object.keys(watchedDirectories);
+                    // Try to find parent paths that are already watched. If found, don't add directory watchers
+                    let watchedParentPaths = watchedPaths.filter(path => fileName.indexOf(path) === 0);
+                    // If adding new watchers, try to find children paths that are already watched. If found, close them. 
+                    if (watchedParentPaths.length === 0) {
+                        let pathToWatch = ts.getDirectoryPath(fileName);
+                        for (let watchedPath in watchedDirectories) {
+                            if (watchedPath.indexOf(pathToWatch) === 0) {
+                                watchedDirectories[watchedPath].close();
+                                delete watchedDirectories[watchedPath];
+                            }
+                        }
+                        watchedDirectories[pathToWatch] = _fs.watch(
+                            pathToWatch, 
+                            (eventName: string, relativeFileName: string) => fileEventHandler(eventName, ts.normalizePath(ts.combinePaths(pathToWatch, relativeFileName)))

```

Comment:
Thing is we don't know what the relative file name is "relative" to. For example, we may have a watcher for `C:\project\test` and later on replaced by a higher level `C:\project` because turns out we need to watch files in that folder as well, so `_fs.watch` returns different relative path for the same file in the `C:\project\test` folder.